[Emails] Improve email management messaging for non owners

AllTerrainDeveloper opened the pull request:
#### Proposed Changes

This change uses the previous changes for the messaging in email comparison, but now in the email management screen.
This email management screen might have 2 different errors.

- When the user is not the owner of the email subscription
- When the user is not the owner of the domain

I've added code to differences them and print the message according to the situation

#### Testing Instructions

Might be a bit difficult to test:

**Scenario 1: Having an email subscription and domain, which belong to your user.**
1. Go to Upgrades -> Emails
2. Chose the right domain (according to the scenario)
3. Check that there is no warning present
4. Check that you are able to add mailboxes

![image](https://user-images.githubusercontent.com/5689927/193028126-6bbfefcc-f13f-46ec-b047-17a03b000123.png)


**Scenario 2: Having an email & domain that does not belong to you.**

You can use Calypso to transfer your domain to another user. This will move the domain and the email subscription
1. Go to Upgrades -> Emails
2. Chose the right domain (according to the scenario)
3. Check that you are presented with a warning
4. Check that you are not able to add mailboxes

![image](https://user-images.githubusercontent.com/5689927/193280162-633253ae-aa84-411f-89c1-334ddc3a937d.png)

**Scenario 3: Having an email (yours) whose domain does not belong to you**

You can use Calypso to transfer your domain to another user. This will move the domain and the email subscription. Also to achieve this situation, please refer to this p2: pbLl1t-LC-p2  and transfer the professional email subscription to you.
1. Go to Upgrades -> Emails
2. Chose the right domain (according to the scenario)
3. Check that you are presented with a warning
4. Check that you are not able to add mailboxes

![image](https://user-images.githubusercontent.com/5689927/193027809-62c3b96b-e283-4120-b1af-ed6462b13d1f.png)

**Scenario 4: Remove all mailboxes form the email account and assert that generic notice with CTA appears.**
![image](https://user-images.githubusercontent.com/5689927/193593669-e52c5f2f-ea99-412c-9c4b-1fea898f312a.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

AllTerrainDeveloper committed:
Emails: Add informative message and actions when user can't add email to the selected domain

AllTerrainDeveloper committed:
Emails: Disable select buttons, gather information for the domain owner and use it in the new link buttons

AllTerrainDeveloper committed:
Emails: Amend Pull Request suggestions
 - Move logic to extract the domain user name
 - Explicit function description to ease the location of the event logged in Tracks
 - Change the style class name & padding for the notice.
 - Extract the function that builds the URL login page with email comparison and user login name

AllTerrainDeveloper committed:
[Emails] Make the URL relative to the host.

AllTerrainDeveloper committed:
Emails: Improve the final wording, build login query correctly, extract component to do not overflow the main render.

AllTerrainDeveloper committed:
Fix linting errors

AllTerrainDeveloper committed:
Emails: Refactor the owner notice into a new component

AllTerrainDeveloper committed:
Emails: Remove test flag

AllTerrainDeveloper committed:
Emails: initial commit

AllTerrainDeveloper committed:
Emails: Address feedback from Pull Request. Added new translations and fixed URL in login action

AllTerrainDeveloper committed:
Emails: Code improvements, rel = external on login link, moved function queryString to a higher scope.

AllTerrainDeveloper committed:
Differentiate between the two codes sent by the backend

AllTerrainDeveloper committed:
Fix linting error

wongasy commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -19,18 +20,47 @@ type EmailPlanWarningsProps = {
+				);
+			case WARNING_CODE_OTHER_USER_OWNS_DOMAIN_SUBSCRIPTION:
+				return (
+					<EmailNonDomainOwnerMessage
```
We should be showing this message regardless of the error. If we're worried about the edge case when the user is a domain owner but not the subscription owner, we may need to update `useDomainOwnerUserName` so that it handles a query for getting email subscription owner as well. However, in reality, the owner of the domain should always be the owner of the email subscription. I'm inclined to not worry about that edge case. @fredrikekelund Thoughts?

AllTerrainDeveloper replied:
Well, it might seem an edge case, but every month a few support cases are indicating that it happens more often than we though :)

wongasy reviewed (changes requested):
@AllTerrainDeveloper I was debugging and added a console log in `useDomainOwnerUserName`, then realized the console log wouldn't stop printing. Looks like `useDomainOwnerUserName` is being called "infinitely", is there any way to avoid that?

AllTerrainDeveloper committed:
Emails: Fix infinite loop with useMemo

AllTerrainDeveloper committed:
Emails: Improve useMemo function

AllTerrainDeveloper committed:
Improve the use of useMemo and remove superfluous code.

AllTerrainDeveloper committed:
Emails: Dynamic redirect based on source context.

AllTerrainDeveloper committed:
Emails: Add another component in order to simplify the code for non domain owners.

AllTerrainDeveloper committed:
Emails: Change the message for email non-owners with a new component

AllTerrainDeveloper committed:
Emails: Add component tracking to the new component.

AllTerrainDeveloper committed:
Emails: Remove un-needed properties, as well as translations not needed

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7557366

Thank you @AllTerrainDeveloper for including a screenshot in the description! This is really helpful for our translators.

wongasy commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -19,19 +22,44 @@ type EmailPlanWarningsProps = {
 	const warning = emailAccount?.warnings?.[ 0 ];
 
+	const cannotAddEmailWarningReason = getCurrentUserCannotAddEmailReason( domain );
+	const cannotAddEmailWarningMessage = cannotAddEmailWarningReason?.message ?? '';
```
Looks like we don't actually need `cannotAddEmailWarningMessage` and can solely rely on `cannotAddEmailWarningCode` in all the checks?

AllTerrainDeveloper replied:
Not really sure, this is how the other messages were being handled, I don't really know if there might be an error code with no message, if that's the case we are going to show a bigger empty space due to the divs being rendered without a message, in my opinion, I prefer keeping the check with the message.

wongasy commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -19,19 +22,44 @@ type EmailPlanWarningsProps = {
 };
 
+const WARNING_CODE_OTHER_USER_OWNS_DOMAIN_SUBSCRIPTION = 'other-user-owns-subscription';
+const WARNING_CODE_OTHER_USER_OWNS_EMAIL = 'other-user-owns-email';
```
Would it make sense to define these constants in `/lib/emails/email-provider-constants.js`?

wongasy commented on client/my-sites/email/email-non-owner-message/index.tsx:
```diff
@@ -0,0 +1,98 @@
+	return (
+		<>
+			<TrackComponentView
+				eventName="calypso_non_email_owner_notice_component_impression"
```
```suggestion
				eventName="calypso_email_providers_nonowner_impression"
```
Same comments from the main PR regarding tracking (https://github.com/Automattic/wp-calypso/pull/68158#discussion_r984688935)

wongasy replied:
Let's rename this event.

wongasy commented on client/my-sites/email/email-non-owner-message/index.tsx:
```diff
@@ -0,0 +1,98 @@
+
+	const onClickLink = ( eventType: 'login' | 'support' ) => {
+		const properties = {
+			action: eventType,
```
```suggestion
			action: eventType,
			source: 'email-management',
```

Let's track the source of the click as well!

wongasy reviewed (changes requested):
Looking good! Just a few minor comments.

AllTerrainDeveloper committed:
Add source as a parameter for Tracks in view component and Track clicks.
Fix async issue happening due to the prop that handles the notice visibility sometimes not fetched yet.

AllTerrainDeveloper committed:
Save the constant on the constants file
Add source in tracking

fredrikekelund commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -21,17 +28,45 @@ type EmailPlanWarningsProps = {
+	const cannotAddEmailWarningMessage = cannotAddEmailWarningReason?.message ?? '';
+	const cannotAddEmailWarningCode = cannotAddEmailWarningReason?.code ?? null;
+
+	const WarningMessageForCode = useMemo( () => {
```
I think we should be able to do this without `useMemo`. Off the top of my head, my preferred solution would probably be to just make `WarningMessageForCode` a regular component and define it outside of the `EmailPlanWarnings` scope (i.e. quite similar to what we had in f774c23bfab20f0ffdbab108db19a1c0b7198b8c, except with a different scope).

@wongasy AFAICT, it's not an issue if `useDomainOwnerUserName` is called repeatedly. `react-query` has internal state management that will only make the required network requests if they aren't already in progress.

fredrikekelund commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -21,17 +28,45 @@ type EmailPlanWarningsProps = {
+							selectedSite={ selectedSite }
+						/>
+					);
+				default:
+					return null;
```
`cannotAddEmailWarningMessage` could contain information about other semi-esoteric states of the domain. I'm wondering if we shouldn't render that message in the default case here.

AllTerrainDeveloper committed:
Missed un-staged change to commit: Add source to props deconstruction.

AllTerrainDeveloper committed:
Extract logic to a new component
Add new styles to the main sheet
Keep showing old notices.

wongasy commented on client/my-sites/email/email-management/home/email-plan-warnings.tsx:
```diff
@@ -21,17 +22,18 @@ type EmailPlanWarningsProps = {
+	const selectedSite = useSelector( getSelectedSite );
+	const selectedSiteSlug = selectedSite?.slug ?? '';
 
 	const warning = emailAccount?.warnings?.[ 0 ];
```
To help distinguish the messaging shown when users cannot make a purchase from the messaging shown when users have need to finish setting up the email/mailboxes, should we rename `warning` to `requiredAction`?

![image](https://user-images.githubusercontent.com/104910361/193648351-51c6e0fa-2806-4e07-b3cf-cf8cffb69aa4.png)

wongasy reviewed (changes requested):
Changes look great! My apologies for requesting a [block of code](https://github.com/Automattic/wp-calypso/commit/75b4a5524e37f0a501b598704e410ec5489f9da9) to be removed from the [main PR](https://github.com/Automattic/wp-calypso/pull/68158), I missed the fact that we **will** be reusing the `EmailNonDomainOwnerMessage` component in email management. 

So we do need this block of code afterall. Let's add it in this PR, so we can merge the other PR which already has the majority of the translation in place.

<img width="785" alt="Screen Shot 2022-10-03 at 2 25 56 PM" src="https://user-images.githubusercontent.com/104910361/193651064-31799ffc-d97a-49cf-b97b-dbd54bb2fa52.png">

I was able to confirm all 4 test scenarios:

* No warning for domain and email owner

<img width="1075" alt="email-owner-no-warning" src="https://user-images.githubusercontent.com/104910361/193651821-47afe299-8f6d-4917-bc2a-b1b2c4fbf7a7.png">

* Warning message for non-email owner

<img width="1009" alt="email-non-owner-warning" src="https://user-images.githubusercontent.com/104910361/193651871-baf6b318-9c30-48de-bc21-202ec122c9fa.png">

* Warning message for non-domain owner (we just need the messaging updated to say "Additional mailboxes..."

<img width="1012" alt="Screen Shot 2022-10-03 at 2 17 18 PM" src="https://user-images.githubusercontent.com/104910361/193651889-8ab3f4a2-d792-49b2-b7a0-a54267659155.png">

* Mailboxes removed

<img width="1071" alt="Screen Shot 2022-10-03 at 2 32 59 PM" src="https://user-images.githubusercontent.com/104910361/193653433-3ab233d0-6075-47a8-9326-8250e84819a9.png">

AllTerrainDeveloper committed:
Emails: Remove the Forward product link at the bottom of the screen when the user is not an owner.

AllTerrainDeveloper committed:
Emails: Change event name in the TrackComponentView, rename warning to requiredAction

wongasy commented:
@AllTerrainDeveloper Thanks for addressing the feedbacks! There's [one last thing that we need to update](https://github.com/Automattic/wp-calypso/pull/68435#pullrequestreview-1128802869), which is to re-add the "Additional mailboxes..." messaging to the non-domain-owner messaging. I had asked for it to be moved in a previous Slack conversation, it was an oversight on my end that the `EmailNonDomainOwnerMessage` component is re-used in Email Management and that messaging is indeed needed.

wongasy commented on client/lib/emails/email-provider-constants.js:
```diff
@@ -8,3 +8,6 @@ export const EMAIL_USER_ROLE_ADMIN = 'admin';
 export const EMAIL_WARNING_SLUG_UNUSED_MAILBOXES = 'unused_mailboxes';
 export const EMAIL_WARNING_SLUG_UNVERIFIED_FORWARDS = 'unverified_forwards';
+export const EMAIL_WARNING_CODE_OTHER_USER_OWNS_DOMAIN_SUBSCRIPTION =
+	'other-user-owns-subscription';
```
```suggestion
	'other-user-owns-domain';
```

AllTerrainDeveloper replied:
That's not our choice, that's the constant we receive through the domains endpoint :)

wongasy commented on client/lib/emails/email-provider-constants.js:
```diff
@@ -8,3 +8,6 @@ export const EMAIL_USER_ROLE_ADMIN = 'admin';
 export const EMAIL_WARNING_SLUG_GOOGLE_ACCOUNT_TOS = 'google_pending_tos_acceptance';
 export const EMAIL_WARNING_SLUG_UNUSED_MAILBOXES = 'unused_mailboxes';
 export const EMAIL_WARNING_SLUG_UNVERIFIED_FORWARDS = 'unverified_forwards';
+export const EMAIL_WARNING_CODE_OTHER_USER_OWNS_DOMAIN_SUBSCRIPTION =
```
```suggestion
export const EMAIL_WARNING_CODE_OTHER_USER_OWNS_DOMAIN =
```

wongasy commented on client/my-sites/email/email-management/home/email-plan-warning-notice.tsx:
```diff
@@ -0,0 +1,53 @@
+					source="email-management"
+				/>
+			);
+		case EMAIL_WARNING_CODE_OTHER_USER_OWNS_DOMAIN_SUBSCRIPTION:
```
```suggestion
		case EMAIL_WARNING_CODE_OTHER_USER_OWNS_DOMAIN:
```

wongasy reviewed (approved):
LGTM! I just have a final minor comment around the constant name and value.

Also, please ignore my [previous comment](https://github.com/Automattic/wp-calypso/pull/68435#pullrequestreview-1128802869) regarding add the "Additional mailboxes..." messaging to the non-domain-owner component. It appears that users will always get the non-email-owner messaging when they don't own the email subscription. In the case where the is an email owner, but not a domain owner, we should be showing another message, which I have added in a [new PR](https://github.com/Automattic/wp-calypso/pull/68601).

stephanethomas commented on client/components/data/query-email-owner-username/index.ts:
```diff
@@ -0,0 +1,52 @@
+
+type User = {
+	ID: number;
+	linked_user_ID: number;
```
What is a _linked user_? Could we find a more explicit name which better conveys what that is?

AllTerrainDeveloper replied:
I'm using this type in order to TypeScript don't complain when calling the reactQuery method which is in JS (not typed). It's not my choice, its the value we are receiving thought the users endpoint API.

If its a Jetpack site, the ID is overridden and the linked_user_ID is the real WordPress user. This gave us a big headache.

stephanethomas replied:
Could we then document it in some ways, either here or when we call `teams.users?.find`?

stephanethomas commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -107,9 +120,14 @@ export const EmailNonDomainOwnerMessage = ( props: EmailNonDomainOwnerMessagePro
-			<PromoCard className="email-non-domain-owner-message__non-owner-notice">
-				<p>{ reasonText }</p>
-			</PromoCard>
+			{ usePromoCard && (
+				<PromoCard className="email-non-domain-owner-message__non-owner-notice">
+					<p>{ reasonText }</p>
+				</PromoCard>
+			) }
+			{ ! usePromoCard && (
+				<p className={ 'email-non-domain-owner-message__non-owner-message' }>{ reasonText }</p>
+			) }
```
That's only a suggestion but I find the following form syntax clear and shorter:

```suggestion
			{ usePromoCard ? (
				<PromoCard className="email-non-domain-owner-message__non-owner-notice">
					<p>{ reasonText }</p>
				</PromoCard>
			) : (
				<p className={ 'email-non-domain-owner-message__non-owner-message' }>{ reasonText }</p>
			) }
```

stephanethomas commented on client/my-sites/email/email-non-owner-message/index.tsx:
```diff
@@ -0,0 +1,100 @@
+		components: {
+			contactSupportLink: (
+				<a
+					href="https://wordpress.com/help/contact"
```
We should use the `CALYPSO_CONTACT` constant rather than hardcoding that url.

AllTerrainDeveloper committed:
Use CALYPSO_CONTACT instead of hardcoding the URL for contact.

AllTerrainDeveloper committed:
Emails: Add comments to clarify the use of linked_user_ID

wongasy committed:
Show email forwarding link for non-domain-owner admins

AllTerrainDeveloper committed:
Add context to Tracks

AllTerrainDeveloper committed:
Better readability when handling usePromoCard flag

AllTerrainDeveloper committed:
Missed merge conflict during rebase

olaseni commented on client/components/data/query-email-owner-username/index.ts:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	const teams = data as InfiniteData< UsersData > & UsersData;
+	const ownerUser = teams.users?.find(
```
I think we should have an equivalent comment here, like we added in the other file.

olaseni commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -6,21 +6,26 @@ import PromoCard from 'calypso/components/promo-section/promo-card';
 	domain?: ResponseDomain;
 	selectedSite?: SiteDetails | null;
-	source: string;
+	source: 'email-comparison' | 'email-management';
```
This feels like a good place to introduce some string constants.

olaseni reviewed (commented):
I have got a few non-blocking suggestions, but mostly it looks good to me.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7557366).
